File: 20231017/python/hyeonjeong.py
from collections import deque

N, H = map(int, input().split())

moves = {i: i for i in range(1, 101)}  # 뱀과 사다리
visited = [0 for _ in range(101)]
for _ in range(N):
    s, e = map(int, input().split())
    moves[s] = e
for _ in range(H):
    s, e = map(int, input().split())
    moves[s] = e

queue = deque([(1, 0)])
while queue:
    cur, cnt = queue.popleft()
    cur = moves[cur]

    if visited[cur]:  # 전에 왔던 곳은 pass
        continue
    visited[cur] = cnt

    if cur == 100:  # 가장 빨리 100 도착하는 거 출력
        print(cnt)
        break

    for i in range(1, 7):
        if cur + i > 100:  # 100을 넘어가면 pass
            break
        queue.append((cur + i, cnt + 1))

# 34052 kb 68 ms

File: 20231016/java/hyeonjeong.java
from collections import deque
import sys

def solution(n, start, end, roads, traps):
    routes = [[0 for _ in range(n + 1)] for _ in range(n + 1)]
    for s, e, c in roads:
        if not routes[s][e] or routes[s][e] > c:
            routes[s][e] = c

    # flipped[start] = -1이면 안뒤집은 상태, 1이면 뒤집은 상태
    flipped = [-1 for _ in range(n + 1)]
    isTrap = [False for _ in range(n+1)]
    for t in traps:
        isTrap[t] = True

    # 위치, 시간, flipped 배열
    queue = deque([(start, 0, flipped)])
    visited = [[-1 for _ in range(n + 1)] for _ in range(n + 1)]
    
    answer = sys.maxsize
    while queue:
        pos, time, flip = queue.popleft()

        if time > answer:
            continue
        
        # 끝에 도착했고, 최단 시간이면 update
        if pos == end:
            if time < answer:
                answer = time
            else:
                continue
        
        # 지금 있는 위치에 trap 있으면 뒤집기
        if isTrap[pos]:
            flip[pos] *= -1
        
        for i in range(1, n+1):
            # 그대로
            if flip[pos] == flip[i]:
                if routes[pos][i] and visited[pos][i] != time:
                    queue.append((i, time + routes[pos][i], flip))
                    visited[pos][i] = time
            # 뒤집어 있음
            else:
                if routes[i][pos] and visited[pos][i] != time:
                    queue.append((i, time + routes[i][pos], flip))
                    visited[i][pos] = time

    return answer

# 5, 26 시간 초과
# 6, 7, 8, 18, 19, 24 실패
